torch/_inductor/fx_passes/fuse_attention.py

```diff
@@ -321,6 +321,69 @@ def _sfdp_replacement_13(query, key, value, dropout_p):
     ).squeeze(0)
 
 
+def _sfdp_pattern_14(query, key, value, attn_mask, inv_scale):
+    # for BertLarge
+    q = query.permute(0, 2, 1, 3)
+    k = key.permute(0, 2, 1, 3)
+    v = value.permute(0, 2, 1, 3)
```

Comment:
Question: Why do we need these transposes to be included in the pattern? They seem to appear in both the pattern and the replacement, so I think ignoring them should give a more general pattern.

Reply:
You are right! Thanks and modified.

Reply:
@peterbell10 Confirmed again, these transposes are necessary for the pattern match. The `clone` would disappear without transposes, leading to the pattern match failure.

Target graph:
```
def forward(self, query_1, key_1, value_1, attn_mask_1, inv_scale_1):
    permute = torch.ops.aten.permute.default(query_1, [0, 2, 1, 3]);  query_1 = None
    permute_1 = torch.ops.aten.permute.default(key_1, [0, 2, 1, 3]);  key_1 = None
    permute_2 = torch.ops.aten.permute.default(value_1, [0, 2, 1, 3]);  value_1 = None
    permute_3 = torch.ops.aten.permute.default(permute_1, [0, 1, 3, 2]);  permute_1 = None
    expand = torch.ops.aten.expand.default(permute, [2, 8, 4, 16]);  permute = None
    **clone = torch.ops.aten.clone.default(expand, memory_format = torch.contiguous_format);  expand = None**
    view = torch.ops.aten.view.default(clone, [16, 4, 16]);  clone = None
    expand_1 = torch.ops.aten.expand.default(permute_3, [2, 8, 16, 4]);  permute_3 = None
    clone_1 = torch.ops.aten.clone.default(expand_1, memory_format = torch.contiguous_format);  expand_1 = None
    view_1 = torch.ops.aten.view.default(clone_1, [16, 16, 4]);  clone_1 = None
    bmm = torch.ops.aten.bmm.default(view, view_1);  view = view_1 = None
    view_2 = torch.ops.aten.view.default(bmm, [2, 8, 4, 4]);  bmm = None
    div = torch.ops.aten.div.Tensor(view_2, inv_scale_1);  view_2 = inv_scale_1 = None
    add = torch.ops.aten.add.Tensor(div, attn_mask_1);  div = attn_mask_1 = None
    convert_element_type_2 = torch.ops.prims.convert_element_type.default(add, torch.float32);  add = None
    amax = torch.ops.aten.amax.default(convert_element_type_2, [-1], True)
    sub = torch.ops.aten.sub.Tensor(convert_element_type_2, amax);  convert_element_type_2 = amax = None
    exp = torch.ops.aten.exp.default(sub);  sub = None
    sum_1 = torch.ops.aten.sum.dim_IntList(exp, [-1], True)
    div_1 = torch.ops.aten.div.Tensor(exp, sum_1);  exp = sum_1 = None
    convert_element_type_3 = torch.ops.prims.convert_element_type.default(div_1, torch.float16);  div_1 = None
    expand_2 = torch.ops.aten.expand.default(convert_element_type_3, [2, 8, 4, 4]);  convert_element_type_3 = None
    view_3 = torch.ops.aten.view.default(expand_2, [16, 4, 4]);  expand_2 = None
    expand_3 = torch.ops.aten.expand.default(permute_2, [2, 8, 4, 16]);  permute_2 = None
    clone_2 = torch.ops.aten.clone.default(expand_3, memory_format = torch.contiguous_format);  expand_3 = None
    view_4 = torch.ops.aten.view.default(clone_2, [16, 4, 16]);  clone_2 = None
    bmm_1 = torch.ops.aten.bmm.default(view_3, view_4);  view_3 = view_4 = None
    view_5 = torch.ops.aten.view.default(bmm_1, [2, 8, 4, 16]);  bmm_1 = None
    return view_5
```

Pattern graph with transposes:
```
def forward(self, query_1, key_1, value_1, attn_mask_1, inv_scale_1):
    permute = torch.ops.aten.permute.default(query_1, [0, 2, 1, 3]);  query_1 = None
    permute_1 = torch.ops.aten.permute.default(key_1, [0, 2, 1, 3]);  key_1 = None
    permute_2 = torch.ops.aten.permute.default(value_1, [0, 2, 1, 3]);  value_1 = None
    permute_3 = torch.ops.aten.permute.default(permute_1, [0, 1, 3, 2]);  permute_1 = None
    expand = torch.ops.aten.expand.default(permute, [2, 8, 4, 16]);  permute = None
    **clone = torch.ops.aten.clone.default(expand, memory_format = torch.contiguous_format);  expand = None**
    view = torch.ops.aten.view.default(clone, [16, 4, 16]);  clone = None
    expand_1 = torch.ops.aten.expand.default(permute_3, [2, 8, 16, 4]);  permute_3 = None
    clone_1 = torch.ops.aten.clone.default(expand_1, memory_format = torch.contiguous_format);  expand_1 = None
    view_1 = torch.ops.aten.view.default(clone_1, [16, 16, 4]);  clone_1 = None
    bmm = torch.ops.aten.bmm.default(view, view_1);  view = view_1 = None
    view_2 = torch.ops.aten.view.default(bmm, [2, 8, 4, 4]);  bmm = None
    div = torch.ops.aten.div.Tensor(view_2, inv_scale_1);  view_2 = inv_scale_1 = None
    add = torch.ops.aten.add.Tensor(div, attn_mask_1);  div = attn_mask_1 = None
    convert_element_type_2 = torch.ops.prims.convert_element_type.default(add, torch.float32);  add = None
    amax = torch.ops.aten.amax.default(convert_element_type_2, [-1], True)
    sub = torch.ops.aten.sub.Tensor(convert_element_type_2, amax);  convert_element_type_2 = amax = None
    exp = torch.ops.aten.exp.default(sub);  sub = None
    sum_1 = torch.ops.aten.sum.dim_IntList(exp, [-1], True)
    div_1 = torch.ops.aten.div.Tensor(exp, sum_1);  exp = sum_1 = None
    convert_element_type_3 = torch.ops.prims.convert_element_type.default(div_1, torch.float16);  div_1 = None
    expand_2 = torch.ops.aten.expand.default(convert_element_type_3, [2, 8, 4, 4]);  convert_element_type_3 = None
    view_3 = torch.ops.aten.view.default(expand_2, [16, 4, 4]);  expand_2 = None
    expand_3 = torch.ops.aten.expand.default(permute_2, [2, 8, 4, 16]);  permute_2 = None
    clone_2 = torch.ops.aten.clone.default(expand_3, memory_format = torch.contiguous_format);  expand_3 = None
    view_4 = torch.ops.aten.view.default(clone_2, [16, 4, 16]);  clone_2 = None
    bmm_1 = torch.ops.aten.bmm.default(view_3, view_4);  view_3 = view_4 = None
    view_5 = torch.ops.aten.view.default(bmm_1, [2, 8, 4, 16]);  bmm_1 = None
    return view_5
```

Pattern graph w/o transposes:
```
def forward(self, query_1, key_1, value_1, attn_mask_1, inv_scale_1):
    permute = torch.ops.aten.permute.default(key_1, [0, 1, 3, 2]);  key_1 = None
    expand = torch.ops.aten.expand.default(query_1, [2, 4, 8, 16]);  query_1 = None
    view = torch.ops.aten.view.default(expand, [8, 8, 16]);  expand = None
    expand_1 = torch.ops.aten.expand.default(permute, [2, 4, 16, 8]);  permute = None
    view_1 = torch.ops.aten.view.default(expand_1, [8, 16, 8]);  expand_1 = None
    bmm = torch.ops.aten.bmm.default(view, view_1);  view = view_1 = None
    view_2 = torch.ops.aten.view.default(bmm, [2, 4, 8, 8]);  bmm = None
    div = torch.ops.aten.div.Tensor(view_2, inv_scale_1);  view_2 = inv_scale_1 = None
    add = torch.ops.aten.add.Tensor(div, attn_mask_1);  div = attn_mask_1 = None
    convert_element_type_2 = torch.ops.prims.convert_element_type.default(add, torch.float32);  add = None
    amax = torch.ops.aten.amax.default(convert_element_type_2, [-1], True)
    sub = torch.ops.aten.sub.Tensor(convert_element_type_2, amax);  convert_element_type_2 = amax = None
    exp = torch.ops.aten.exp.default(sub);  sub = None
    sum_1 = torch.ops.aten.sum.dim_IntList(exp, [-1], True)
    div_1 = torch.ops.aten.div.Tensor(exp, sum_1);  exp = sum_1 = None
    convert_element_type_3 = torch.ops.prims.convert_element_type.default(div_1, torch.float16);  div_1 = None
    expand_2 = torch.ops.aten.expand.default(convert_element_type_3, [2, 4, 8, 8]);  convert_element_type_3 = None
    view_3 = torch.ops.aten.view.default(expand_2, [8, 8, 8]);  expand_2 = None
    expand_3 = torch.ops.aten.expand.default(value_1, [2, 4, 8, 16]);  value_1 = None
    view_4 = torch.ops.aten.view.default(expand_3, [8, 8, 16]);  expand_3 = None
    bmm_1 = torch.ops.aten.bmm.default(view_3, view_4);  view_3 = view_4 = None
    view_5 = torch.ops.aten.view.default(bmm_1, [2, 4, 8, 16]);  bmm_1 = None
    return view_5
```

Reply:
That's unfortunate. I do wonder if we could update the pattern matcher to not be sensitive to these additional clones, but in the mean time I think this is okay. Probably deserves a comment on the pattern though.

Reply:
Thanks! The comment was added. Please review again :)